streamer/src/nonblocking/quic.rs

```diff
@@ -281,30 +281,61 @@ fn get_connection_stake(
-                    peer_stake, total_stake,
-                );
+pub struct UniStreamQosUtil {}
 
-                QUIC_MIN_STAKED_CONCURRENT_STREAMS
-            } else {
-                let delta = (QUIC_TOTAL_STAKED_CONCURRENT_STREAMS
-                    - QUIC_MIN_STAKED_CONCURRENT_STREAMS) as f64;
-
-                (((peer_stake as f64 / total_stake as f64) * delta) as usize
-                    + QUIC_MIN_STAKED_CONCURRENT_STREAMS)
-                    .clamp(
-                        QUIC_MIN_STAKED_CONCURRENT_STREAMS,
-                        QUIC_MAX_STAKED_CONCURRENT_STREAMS,
-                    )
+impl UniStreamQosUtil {
+    /// Calculate the maximum allowed uni streams based on stake information.
+    pub fn compute_max_allowed_uni_streams(
+        peer_type: ConnectionPeerType,
+        total_stake: u64,
+    ) -> usize {
+        match peer_type {
+            ConnectionPeerType::Staked(peer_stake) => {
+                // No checked math for f64 type. So let's explicitly check for 0 here
+                if total_stake == 0 || peer_stake > total_stake {
+                    warn!(
+                        "Invalid stake values: peer_stake: {:?}, total_stake: {:?}",
+                        peer_stake, total_stake,
+                    );
+
+                    QUIC_MIN_STAKED_CONCURRENT_STREAMS
+                } else {
+                    let delta = (QUIC_TOTAL_STAKED_CONCURRENT_STREAMS
+                        - QUIC_MIN_STAKED_CONCURRENT_STREAMS)
+                        as f64;
+
+                    (((peer_stake as f64 / total_stake as f64) * delta) as usize
+                        + QUIC_MIN_STAKED_CONCURRENT_STREAMS)
+                        .clamp(
+                            QUIC_MIN_STAKED_CONCURRENT_STREAMS,
+                            QUIC_MAX_STAKED_CONCURRENT_STREAMS,
+                        )
+                }
             }
+            ConnectionPeerType::Unstaked => QUIC_MAX_UNSTAKED_CONCURRENT_STREAMS,
         }
-        ConnectionPeerType::Unstaked => QUIC_MAX_UNSTAKED_CONCURRENT_STREAMS,
+    }
+
+    /// Given the load EMA, derive the streams per throttle window.
+    /// Do not allow concurrent streams more than the max streams per throttle window.
+    pub(crate) fn max_allowed_uni_streams_per_throttling_interval_ema(
+        ema: &StakedStreamLoadEMA,
+        peer_type: ConnectionPeerType,
+        total_stake: u64,
+    ) -> u64 {
+        let max_streams_per_throttle_window =
+            ema.available_load_capacity_in_throttling_duration(peer_type, total_stake);
+        (UniStreamQosUtil::compute_max_allowed_uni_streams(peer_type, total_stake) as u64)
+            .min(max_streams_per_throttle_window)
+    }
+
+    /// Given the max_streams_per_throttling_interval, derive the streams per throttle window.
+    /// Do not allow concurrent streams more than the max streams per throttle window.
+    pub fn max_concurrent_uni_streams_per_throttling_interval(
```

Comment:
Do we really need a function for this? It's just a wrapper on `min`. Why not directly use `min()` where we are calling this?

Reply:
I think this makes the design goal more explicit and easier to do test